### DISCUSSION TITLE
'locales' directory not copied to bin during generation (.net core, multi-target, RuntimeIdentifier, CefSharp package reference in class library)

### DISCUSSION THREAD

**Discussion by JPDelprat** [cefsharp/CefSharp#4207] [General]
Hello,

For clarity, and because I found the solution to my problem, here is a synthesis of my messages, which replaces other ones, in case someone is interested.

My use case:
- .net 6.0 (.net core)
- A main project (windows executable) which references a child project (class library), which in turn references CefSharp.WinForms.NETCore package,
- both projects are multiple targets
- RuntimeIdentifier is defined on main project.

Problem:
- CefSharp 'locales' directory is not copied to the bin directory of my main project during generation

Solution:
- in the class library, also reference 'CefSharp.Common.NETCore'  package ; it shouldn't be needed since it is a dependency of CefSharp.WinForms.NETCore, but this solves the problem

**Comment by JPDelprat** [comment]
Solution found. First message updated with all information needed.

**Reply by amaitland** [reply]
> Problem:
> CefSharp 'locales' directory is not copied to the bin directory of my main project during generation

I don't believe there is a problem when not multi targeting. For me this is a bug in `VS/MSBuild` and needs to be fixed by `MS`.

> Solution:
> in the class library, also reference 'CefSharp.Common.NETCore' package ; it shouldn't be needed since it is a dependency of CefSharp.WinForms.NETCore, but this solves the problem

This shouldn't be necessary and I believe this is a bug in `VS/MSBuild`.

- The `.csproj.nuget.g.props` file (in the `obj` folder) is missing a reference to `chromiumembeddedframework.runtime.props`
- The `.csproj.nuget.g.targets` file (in the `obj` folder) file is missing a reference to `CefSharp.Common.NETCore.targets`

**UPDATE** It doesn't look like any of the props or targets for dependant packages are added.

I'm pretty sure https://github.com/dotnet/project-system is the repository where this should be filed, if not they can point you to the correct one. Please post a link back here for reference.

--- 

General guidance on multi targeting, there's a working example at https://github.com/cefsharp/CefSharp.Wpf.HwndHost

- Set [CefSharpBuildAction = NoAction](https://github.com/cefsharp/CefSharp.Wpf.HwndHost/blob/master/CefSharp.Wpf.HwndHost/CefSharp.Wpf.HwndHost.csproj#L20) in `Library` projects so unnecessary files aren't copied.
- Specify a [RuntimeIdentifier](https://github.com/cefsharp/CefSharp.Wpf.HwndHost/blob/master/CefSharp.Wpf.HwndHost.Example/CefSharp.Wpf.HwndHost.Example.csproj#L14) in `WinExe/Exe` projects.
- Add a reference to [CefSharp.Common.NETCore](https://github.com/cefsharp/CefSharp.Wpf.HwndHost/blob/master/CefSharp.Wpf.HwndHost/CefSharp.Wpf.HwndHost.csproj#L30) in `Library` projects in addition to `CefSharp.WinForms.NETCore/CefSharp.Wpf.NETCore/CefSharp.OffScreen.NETCore` (whichever is relevant in your use case).

**Reply by JPDelprat** [reply]
Thanks very much for your answer and explanations.

Just a note though. It seems to me that setting 'CefSharpBuildAction = NoAction' in library is useless for .net core projects (since you do it automatically), and that it prevents the copy of useful files for .net framework projects, so I don't set it.

**Comment by andrewimcclement** [comment]
I am now seeing the same issue _without_ multitargeting - again referencing CefSharp.Common.NETCore in the library project ensured the locales/*.pak files were copied correctly for me, so the workaround is fine. However it still seems strange.
